system/include/emscripten/emmalloc.h

```diff
@@ -0,0 +1,122 @@
+#pragma once
+
+#include <stddef.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// emmalloc: A lightweight web-friendly memory allocator suitable for very small applications.
+// Enable the usage of emmalloc by passing the linker flag -s MALLOC=emmalloc to the application.
+
+// A debug function that dumps the whole structure of malloc internal memory blocks to console.
+// *extremely slow*, use for debugging allocation test cases.
+void emmalloc_dump_memory_regions(void);
+
+// Allocates size bytes with the given pow-2 alignment.
+void *memalign(size_t alignment, size_t size);
+void *emmalloc_memalign(size_t alignment, size_t size);
+void *emscripten_builtin_memalign(size_t alignment, size_t size);
+void *aligned_alloc(size_t alignment, size_t size);
+
+// Allocates size bytes with default alignment (8 bytes)
+void *malloc(size_t size);
+void *emmalloc_malloc(size_t size);
+void *emscripten_builtin_malloc(size_t size);
+
+// Returns the number of bytes that are actually allocated to the given pointer ptr.
+// E.g. due to alignment or size requirements, the actual size of the allocation can be
+// larger than what was requested.
+size_t malloc_usable_size(void *ptr);
+size_t emmalloc_usable_size(void *ptr);
+
+// Frees a memory pointer allocated with any of
+// emmalloc_memalign, emmalloc_malloc, 
+void free(void *ptr);
+void emmalloc_free(void *ptr);
+void emscripten_builtin_free(void *ptr);
+
+// Performs a reallocation of the given memory pointer to a new size. If the memory region
+// pointed by ptr cannot be resized in place, a new memory region will be allocated, old
+// memory copied over, and the old memory area freed. The pointer ptr must have been
+// allocated with one of the emmalloc memory allocation functions (malloc, memalign, ...).
+// If called with size == 0, the pointer ptr is freed, and a null pointer is returned. If
+// called with null ptr, a new pointer is allocated.
+void *realloc(void *ptr, size_t size);
```

Comment:
Do we need to re-declare all the POSIX names here?    Would it be better to use the ones from stdlib.h and only declare the specific emmalloc symbols in this header?

Reply:
I think it's more documentative to re-declare them here. No harm in doing so, in fact, if there ever were a change in signature of one of the names, then this would loudly complain, and it'd trigger to change the signature of the corresponding `emmalloc_` name as well.